### PR TITLE
PPYolo-E 

### PR DESCRIPTION
**Pull request by BloodAxe**

### CODE REVIEW

**Review by shaydeci — CHANGES_REQUESTED** [review]
For the sake of simplicity of the next part- I review most parts of the code, leaving out the loss, which requires me further diving into the PP Yolo E loss implementation for an informative review.

Main points which I addressed in the comments:

- Regarding new code introduced, that might have already an alterantive in SG. namely - nn.Modules: I feel like a lot of the CSP related blocks  (for example SPP which is almost exactly the same) can be reused.
- Docstrings.
- Cleanup leftovers.

Besides that, there are a few comments inline. 

Really appreciate the hard work!

File: src/super_gradients/training/dataloaders/distributed_sampler_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    def __len__(self) -> int:
+        """
+        Returns:
+            int: length of the dataset
+        """
+        return len(self.sampler)
+
+
+class DistributedSamplerWrapper(DistributedSampler):
+    """
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/dataloaders/distributed_sampler_wrapper.py]: Reverted that stuff, since it's no more relevant as we fixed bug with shuffle

File: src/super_gradients/training/models/detection_models/pp_yolo_e/pp_yolo_e.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    def replace_head(self, new_num_classes=None, new_head=None):
+        if new_num_classes is None and new_head is None:
+            raise ValueError("At least one of new_num_classes, new_head must be given to replace output layer.")
+        if new_head is not None:
+            self.head = new_head
+        else:
+            self.head.replace_num_classes(new_num_classes)
+
+
+class PPYoloE_S(PPYoloE):
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/models/detection_models/pp_yolo_e/pp_yolo_e.py:49]: Now the model will load it's corresponding arch params so it should be much more user-friendly

File: src/super_gradients/training/models/detection_models/pp_yolo_e/pan.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+from super_gradients.common.decorators.factory_decorator import resolve_param
+from super_gradients.common.factories.activations_type_factory import ActivationsTypeFactory
+from torch import nn, Tensor
+from super_gradients.training.models.detection_models.csp_resnet import CSPResNetBasicBlock
+from super_gradients.modules import ConvBNAct
+
+__all__ = ["CustomCSPPAN"]
+
+
+class SPP(nn.Module):
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/models/detection_models/pp_yolo_e/pan.py:14]: Yes, these two module looks similar. But they have different name of internal layers. 
This would require us to update pretrained checkpoints on S3 for ppyolo model or retrain them.

SPP block from YoloX could also be refactored to use `ConvBNAct` module, but that would touch YoloX checkpoints as well. I'd rather not open this can of worms for this task, tbh.

File: src/super_gradients/training/datasets/detection_datasets/detection_dataset.py [hunk 4]
```diff
@@ -67,7 +67,7 @@ class DetectionDataset(Dataset):
     def __init__(
         self,
         data_dir: str,
-        input_dim: tuple,
+        input_dim: Optional[Tuple[int, int]],
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/datasets/detection_datasets/detection_dataset.py:70]: In this case - it shouldn't be None by default. This is to keep the BC with existing code and force user to either specify the dim or explicitly set it to None

File: src/super_gradients/training/dataloaders/dataloaders.py [hunk 5]
```diff
@@ -86,6 +89,58 @@ def get_data_loader(config_name, dataset_cls, train, dataset_params=None, datalo
         return dataloader
 
 
+def get_new_data_loader(dataset_cls, dataset_params: Mapping, dataloader_params: Mapping) -> DataLoader:
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/dataloaders/dataloaders.py]: Removed)

**Review by shaydeci — CHANGES_REQUESTED** [review]
Resolve conflicts, then finally LGTM (:

**Comment by BloodAxe** [comment]
@shaydeci conflicts resolved :)

**Review by shaydeci — DISMISSED** [review]
LGTM

**Review by shaydeci — APPROVED** [review]
